### ISSUE TITLE
Perspective Camera from Blender 4x4 World Matrix

### ISSUE THREAD

**Issue by Visuals3D** [facebookresearch/pytorch3d#1105]
Hello, 

I'm trying for days now to import a Blender Camera with 4x4 world Matrix (RT) and calibration Matrix K into a Pytorch3D PerspectiveCamera. 
The coordinate system of Blender is +Z up, -X left, -Y back and Pytorch3D has +Y up, +X left, +Z back. 

So i tried to convert the Blender Matrix in OpenCV format first and use the cameras_from_opencv_projection function:

```
bc = # The 4x4 Matrix from Blender camera.matrix_world

to_opencv = np.array([[1, 0, 0, 0],
                             [0, -1, 0, 0],
                             [0, 0, -1, 0],
                             [0, 0, 0, 1]])

bc =  to_opencv @ bc

print(bc)

t = bc[:3,-1]
r = bc[:3, :3]

cameras = cameras_from_opencv_projection(R=torch.tensor([r], device=self.device).float(),
                                 tvec=torch.tensor([t], device=self.device).float(),
                                 camera_matrix=torch.tensor(K, device=self.device).float())

print("get_camera_center", cameras.get_camera_center())
```

![blender-pytorch3d](https://user-images.githubusercontent.com/74773202/157077884-71d52e07-3e7a-4931-ae29-190060b92e44.png)

I also tried to rotate the blender matrix by -90deg on X and 180deg on Y to transform it to pytorch3d convention and used a PerspectiveCamera. 

In all cases the result was not even close to the blender positions. What i was able to observe was that the camera center changes when the 4x4 Matrix contains rotational changes. Therefore it might be that the rotation is applied after translation by pytorch3d. 
Is that really the behavior of PerspectiveCamera and cameras_from_opencv_projection? 
Or is it just my stupidity or misunderstanding of the api? 

Another thing i don't know how to handle is the fact that Blender cameras looking down the -Z axis (-Y in pytorch3d) by default. And the pytorch3d camera looks in the +Z axis (-Y in Blender). So a 90° rotation around the x axis is needed to get the blender camera on the same up vector as in pytorch3d. 

Anyone else had the need to import Blender cameras into Pytorch3D and can sharea solution for it?

**Comment by Visuals3D** [comment]
Ok i got it now. My guess was right the rotation is applied after translation. 

I decided to use the FoVPerspectiveCamera and convert the 4x4 Matrix directly from Blender to Pytorch3D 

This is done by swapping and flipping y and z. 

The most important thing is to multiply the T vector with R before creating the Camera object to compensate for the rotation after translation order in Pytorch3D. 

After that the camera was positioned and rotated correctly but upside down. To solve that i added a 180 deg rotation around the z axis to the rotation matrix. 

```
swap_y_z = np.array([[1, 0, 0, 0],
                     [0, 0, -1, 0],
                     [0, -1, 0, 0],
                     [0, 0, 0, 1]])

deg180 = np.deg2rad(180)
rot_z = np.array([[np.cos(deg180), -np.sin(deg180), 0],
         [np.sin(deg180), np.cos(deg180), 0],
         [0, 0, 1]])

c2w = swap_y_z @ c2w

t = c2w[:3,-1]  # Extract translation of the camera
r = c2w[:3, :3] @ rot_z # Extract rotation matrix of the camera

t = t @ r # Make rotation local

fov = np.rad2deg(data["camera_angle_x"])
cameras = FoVPerspectiveCameras([r]=torch.tensor(R, device=self.device).float(),
                                 T=torch.tensor([t], device=self.device).float(),
                                 fov =fov)

```

![Imported Cameras](https://user-images.githubusercontent.com/74773202/157766934-b6bc2a4e-0414-4294-9d6b-b5d068fbc5e6.png)

**Comment by gkioxari** [comment]
Wonderful!! Thanks for following up with a solution. I am sure many of our users will appreciate it. I wonder if you also want to make this a discussion topic in the Discussions github page for more visibility!

**Comment by unlugi** [comment]
Hello, thank you for your help @Visuals3D.
I used your approach: extracted camera.matrix_world as is. Imported it for pytorch3d. But when I render my mesh, the results is flipped about the y axis. Do you know what might be the reason? Thank you

**Comment by akanimax** [comment]
> Ok i got it now. My guess was right the rotation is applied after translation.
> 
> I decided to use the FoVPerspectiveCamera and convert the 4x4 Matrix directly from Blender to Pytorch3D
> 
> This is done by swapping and flipping y and z.
> 
> The most important thing is to multiply the T vector with R before creating the Camera object to compensate for the rotation after translation order in Pytorch3D.
> 
> After that the camera was positioned and rotated correctly but upside down. To solve that i added a 180 deg rotation around the z axis to the rotation matrix.
> 
> ```
> swap_y_z = np.array([[1, 0, 0, 0],
>                      [0, 0, -1, 0],
>                      [0, -1, 0, 0],
>                      [0, 0, 0, 1]])
> 
> deg180 = np.deg2rad(180)
> rot_z = np.array([[np.cos(deg180), -np.sin(deg180), 0],
>          [np.sin(deg180), np.cos(deg180), 0],
>          [0, 0, 1]])
> 
> c2w = swap_y_z @ c2w
> 
> t = c2w[:3,-1]  # Extract translation of the camera
> r = c2w[:3, :3] @ rot_z # Extract rotation matrix of the camera
> 
> t = t @ r # Make rotation local
> 
> fov = np.rad2deg(data["camera_angle_x"])
> cameras = FoVPerspectiveCameras([r]=torch.tensor(R, device=self.device).float(),
>                                  T=torch.tensor([t], device=self.device).float(),
>                                  fov =fov)
> ```
> 
> ![Imported Cameras](https://user-images.githubusercontent.com/74773202/157766934-b6bc2a4e-0414-4294-9d6b-b5d068fbc5e6.png)

Unfortunately, this solution is incorrect with regards to the applied rotation around the z-axis which causes the camera to render mirror-images of the scene as pointed out by @unlugi. This breaks nerf related reconstruction tasks completely. 
Instead simply flip the y-axis as follows:

```
t = c2w[:3, -1] # Extract translation of the camera
r = c2w[:3, :3] # Extract rotation matrix of the camera
r[:, 1] *= -1  # flip y axis
t = t @ r  # Make rotation local
```

**Comment by fangtiancheng** [comment]
The following code can obtain the correct results

```python
# run in blender terminal
# cam = D.objects['Camera']
def getCameraExtrinsicForPytorch3d(cam):
    B = cam.matrix_world
    # blender world to OpenGL world
    B2P_R1 = Matrix([[1,0,0],[0,0,1],[0,-1,0]])
    B2P_R2 = Matrix([[-1,0,0],[0,1,0],[0,0,-1]])
    B2P_T  = Matrix([[-1,0,0],[0,0,-1],[0,1,0]])
    R = B2P_R1 @ B.to_3x3() @ B2P_R2
    T = B2P_T @ B.col[3].xyz @ R
    return R, T
```

**Comment by fangtiancheng** [comment]
The coordinate conversion code using numpy can be represented as:
```python
import numpy as np
from typing import Tuple

# pytorch3d to blender
def P2B(R:np.ndarray, T:np.ndarray)->np.ndarray:
    P2B_R1 = np.array([[1,0,0],[0,0,-1],[0,1,0]], dtype=np.float64)
    P2B_R2 = np.array([[-1,0,0],[0,1,0],[0,0,-1]], dtype=np.float64)
    P2B_T  = np.array([[-1,0,0],[0,0,1],[0,-1,0]], dtype=np.float64)
    vec4w  = np.array([[0,0,0,1]], dtype=np.float64)
    Bcol3 = P2B_T @ R @ T
    B3x3  = P2B_R1 @ R @ P2B_R2
    B3x4 = np.concatenate([B3x3, Bcol3[:,None]], axis=1)
    B = np.concatenate([B3x4,vec4w], axis=0)
    return B

# blender to pytorch3d
def B2P(B:np.ndarray)->Tuple[np.ndarray, np.ndarray]:
    B2P_R1 = np.array([[1,0,0],[0,0,1],[0,-1,0]], dtype=np.float64)
    B2P_R2 = np.array([[-1,0,0],[0,1,0],[0,0,-1]], dtype=np.float64)
    B2P_T  = np.array([[-1,0,0],[0,0,-1],[0,1,0]], dtype=np.float64)
    R = B2P_R1 @ B[:3, :3] @ B2P_R2
    T = B2P_T @ B[:3, 3] @ R
    return R, T
```

**Comment by Rlee719** [comment]
`rot_z = torch.tensor([[-1, 0,  0],
                     [0,  -1, 0],
                     [0,  0,  1]]).float().cuda()
rot_x = torch.tensor([[1,  0,  0, 0],
                      [0,  -1, 0, 0],
                      [0,  0, -1, 0], 
                      [0,  0,  0, 1]]).float().cuda()

def convert_coordinate_system_blender_p3d(camera_to_world):
    N = camera_to_world.shape[0]
    camera_to_world[:, :3, :3] = torch.bmm(camera_to_world[:, :3, :3], rot_z.repeat(N, 1, 1))
    camera_to_world[:, :3,  3] = torch.bmm(camera_to_world[:, None,:3, 3], camera_to_world[:, :3, :3]).squeeze(1)
    camera_to_world = torch.bmm(camera_to_world, rot_x.repeat(N, 1, 1))
    return camera_to_world

cameras_batch = PerspectiveCameras(device=device, R=camera_to_world[:,:3,:3], T=camera_to_world[:,:3,3], focal_length=(focal, ))
`

This works for my case, the rendered image from blender and pytorch3d aligns. However, I haven't figured it out why it works yet. I'll be glad if anyone can give a hint.
